design/service-library.md

```diff
@@ -0,0 +1,187 @@
+npm install @microsoft/kiota-authentication-azure --save
+npm install @azure/identity --save
+```
+
+### Importing the right functionalities from the Graph JavaScript Service Library
+
+```typescript
+import { Client } from "@microsoft/microsoft-graph-client";
+import { User, Message, BodyType } from "@microsoft/microsoft-graph-client/models";
+import { AzureIdentityAuthenticationProvider } from "@microsoft/kiota-authentication-azure";
+import { DeviceCodeCredential } from "@azure/identity";
+```
+
+### Setting up the app to work with the Graph JavaScript Service Library
+
+```typescript
+const deviceCodeCredentials = new DeviceCodeCredential({
+	tenantId: "b61f9af1-d6cf-4cc0-a6f6-befb38bc00ed",
+	clientId: "bde251a6-0ef9-42a8-a40b-9ad9bb594b2c",
+});
+
+const scopes = ["User.Read", "Mail.Send"];
+
+const graphClient = Client.init({
+	authenticationTokenProvider: new AzureIdentityAuthenticationProvider(deviceCodeCredentials, scopes),
+});
+```
+
+### Calling Microsoft Graph via the Graph JavaScript Service Library
+
+The core value of the Microsoft Graph JavaScript Service Library is the availability of the chaned request builder pattern (fluent API). This provides developers with typechecking and discoverability when building on Microsoft Graph. This would highlight the full spectrum of capabilities on Microsoft Graph directly in the API Client.
+
+Models are also available in this package and should reflect the underlying version of Graph we are targeting. Developers using our types should do it in a way that will be familiar to how they are used to with other API Clients and APIs. We should provide an easy-to-use model that feels natural in the JavaScript world. This means we should not be forcing developers to use Classes and / or other structures that will make their code less natural to them.
+
+Because of the richness of the Graph APIs, we are making a decision to require developers to selectively import the areas of Graph they care about. This will provide a lightweight bundle size while still offering a good developer experience.
+
+### Selective Imports of the Graph areas needed
+
+To offer a bundle size that fair to the developer, we are using a selective imports strategy. This strategy enables developers to bring only the endpoints they need from Graph. Developers would only see the root paths of Graph when using the API Client by default, and would need to import additional path with selective imports.
+
+#### Using the service library without any selective imports
+
+> ❌ **Fails** <br />
+> This example does not work because `usersById().directReports` is not available at the root of the Graph Client
+
+```typescript
+const directReports = await graphClient.usersById("ab677c1b-50c6-4013-aa9f-792ed42c59c8").directReports.get();
+```
+
+> ✅ **Succeeds** <br />
+> This example succeeds because `me` is at the root of the Graph Client
+
+```typescript
+const me = await graphClient.me.get();
```

Comment:
I don't think we should make a different between first level path segments and the rest. This introduces a inconsistency of experience.

Reply:
I think we're losing a lot of discoverability if the bare API Client has no references to anything. We're not talking about the self-serve option here, we're expecting the "full" API available. Thoughts @darrelmiller @gavinbarron @nikithauc

Reply:
We're in the realm of trade-offs here. I agree that import hurts discoverability a lot, the benefit is that by being explicit which children of the root node you want to pull in then the produced app bundle becomes much smaller as it's not including root nodes your app doesn't want.

My instinct here is to try the more discoverable approach in the first instance and evaluate the produced bundle size to see if we need to optimize for size over discoverability.